cdc/log.cc

```diff
@@ -538,13 +538,119 @@ class transformer final {
             _cdc_ttl_opt = std::chrono::seconds(_schema->cdc_options().ttl());
         }
     }
+    static size_t collection_size(const bytes_opt& bo) {
+        if (bo) {
+            bytes_view bv(*bo);
+            return read_collection_size(bv, cql_serialization_format::internal());
+        }
+        return 0;
+    }
+    template<typename Func>
+    static void map_for_each(const bytes_opt& bo, Func&& f) {
+        if (bo) {
+            bytes_view bv(*bo);
+            auto size = read_collection_size(bv, cql_serialization_format::internal());
+            while (size--) {
+                auto k = read_collection_value(bv, cql_serialization_format::internal());
+                auto v = read_collection_value(bv, cql_serialization_format::internal());
+                f(k, v);
+            }
+        }
+    }
+    template<typename Func>
+    static void set_for_each(const bytes_opt& bo, Func&& f) {
+        if (bo) {
+            bytes_view bv(*bo);
+            auto size = read_collection_size(bv, cql_serialization_format::internal());
+            while (size--) {
+                f(read_collection_value(bv, cql_serialization_format::internal()));
+            }
+        }
+    }
+    template<typename Func>
+    static void udt_for_each(const bytes_opt& bo, Func&& f) {
+        if (bo) {
+            bytes_view bv(*bo);
+            std::for_each(tuple_deserializing_iterator::start(bv), tuple_deserializing_iterator::finish(bv), std::forward<Func>(f));
+        }
+    }
+    static bytes merge(const collection_type_impl& ctype, const bytes_opt& prev, const bytes_opt& next, const bytes_opt& deleted) {
+        std::vector<std::pair<bytes_view, bytes_view>> res;
+        res.reserve(collection_size(prev) + collection_size(next));
+        map_for_each(prev, [&res](bytes_view k, bytes_view v) {
+            res.emplace_back(k, v);
+        });
+        auto type = ctype.name_comparator();
+        auto cmp = [&type = *type](std::pair<bytes_view, bytes_view>& p, const bytes_view& k) {
+            return type.compare(p.first, k) < 0;
+        };
+        map_for_each(next, [&res, &cmp](bytes_view k, bytes_view v) {
+            auto i = std::lower_bound(res.begin(), res.end(), k, cmp);
+            if (i != res.end() && i->first == k) {
+                i->second = v;
+            } else {
+                res.emplace(i, k, v);
```

Comment:
This is still sad. Maybe we could collect those elements in another vector and at the end do std::merge of res and this new vector that keeps elements which live only in next?

Reply:
Why would that be better? Again, in most cases "next" will be small-ish, and merge would not really have better performance (barring a few memmoves), but we would instead need two temporaries more (one for dest, one for tmp source).

Reply:
I agree that next should be small but the problem is that complexity depends both on the size of next and the size of prev. If every element from next gets inserted before every element in prev, we will be moving elements size(next) * size(prev) times. I agree that std::merge with such temporary won't be much better though - just size(next) times faster. I think the best solution would be to implement a special iterator that computes elements of map from bytes and then use such iterators with std::merge. The question is whether you would like to do it yourself or should we do it in the following PR?

Reply:
say `prev` has `n` elements, `next` has `m` elements
for the first element in `prev`, this `emplace` has to shift `n` elements to the right in the worst case, `n/2` on average
for the second element, it has to shift `n + 1` elements in the worst case, `(n+1)/2` on average
...
for the `k`th element, it has to shift `n+k` elements in the worst case, `(n+k)/2` on average
and so on until `m`

so on average we have to perform `1/2 * ((n+1) + (n + 2) + ... + (n + m))` operations, in the worst case remove the `1/2`

so in total, the average complexity and the worst case complexity of this algorithm is `O((n+m) * m)`

Reply:
we can solve the problem using `std::merge`:
1. compute a sorted vector of elements from `prev`
2. compute a sorted vector of elements from `next`
3. merge the vectors using the merge algorithm, which has `O((n+m) * log (n+m))` complexity

Reply:
> say `prev` has `n` elements, `next` has `m` elements
> for the first element in `prev`, this `emplace` has to shift `n` elements to the right in the worst case, `n/2` on average
> for the second element, it has to shift `n + 1` elements in the worst case, `(n+1)/2` on average
> ...
> for the `k`th element, it has to shift `n+k` elements in the worst case, `(n+k)/2` on average
> and so on until `m`
> 
> so on average we have to perform `1/2 * ((n+1) + (n + 2) + ... + (n + m))` operations, in the worst case remove the `1/2`
> 
> so in total, the average complexity and the worst case complexity of this algorithm is `O((n+m) * m)`

This is not true because both maps are sorted. You're at most moving n*m times IMO

Reply:
> we can solve the problem using `std::merge`:
> 
> 1. compute a sorted vector of elements from `prev`
> 2. compute a sorted vector of elements from `next`
> 3. merge the vectors using the merge algorithm, which has `O((n+m) * log (n+m))` worst case complexity

This is is worse than having special iterator that extracts maps from bytes and use merge on those iterators. It requires 3 allocations instead of 1 and using 2 * (n + m) memory instead of just (n+m).

Reply:
The "sorted vector" is conceptual, it doesn't have to reside in memory all at once of course. You're right, an iterator can be used because we just need to walk over the vector once

Reply:
But the very least thing we need to do is reduce the complexity, **then** we can fight for improving the constants (2*x instead of x etc.)

Reply:
Argh, stupid error.
The merge algorithm has `O(n+m)` complexity, not `O((n+m) log (n+m))`.
So it will be linear because our maps are already sorted :)

Reply:
> This is not true because both maps are sorted. You're at most moving n*m times IMO

Right. You never shift elements from the `next` map, because you walk it in sorted order. So on average it's `n/2` elements shifted for every element in `next`, i.e. `n/2 * m` shifts total on average. `n * m` shifts total in the worst case.

Reply:
@elcallio Do you want to do this or should Kamil or me do it?

Reply:
Now iterator-based.